### ISSUE TITLE
[nri-metadata-injection] Webhook timeout can block Digital Ocean cluster upgrades

### ISSUE THREAD

**Issue by ianonavy** [newrelic/helm-charts#788]
**Bug description**

Digital Ocean's [clusterlint](https://github.com/digitalocean/clusterlint) tool recommends setting [webhook timeouts](https://docs.digitalocean.com/products/kubernetes/resources/clusterlint-errors/#admission-controller-webhook-timeout) to be 1 to 29 seconds in order to avoid blocking automated cluster upgrades. The [current default](https://github.com/newrelic/helm-charts/blob/520c3e6039ea67ae9099724191531bab37142f43/charts/nri-metadata-injection/values.yaml#L84) is 30 seconds.

**Version of Helm and Kubernetes**

I'm using flux2 v0.17.2 on Kubernetes 1.20.15-do.1, but I'm not sure the versions matter as this problem affects Digital Ocean Kubernetes clusters of all versions.

**Which chart?**

nri-metadata-injection v1.1.1 (technically, I am using bundle version 3.5.0)

**What happened?**

I ran the "operational readiness check" feature on my production Digital Ocean cluster in preparation of a Kubernetes version upgrade after installing the NRI bundle chart with mostly default values, and I got this error: 

>Mutating webhook with a TimeoutSeconds value smaller than 1 second or greater than 29 seconds will block upgrades.

<details><summary>Screenshot:</summary>
<img width="787" alt="screenshot of a modal in the operational readiness check feature of the Digital Ocean console, text reads: Mutating webhook with a TimeoutSeconds value smaller than 1 second or greater than 29 seconds will block upgrades." src="https://user-images.githubusercontent.com/790279/164369213-e547b59c-2594-404e-b161-5116b04654f2.png">
</details>

The other issue in the report is the use of [hostpath volumes](https://docs.digitalocean.com/products/kubernetes/resources/clusterlint-errors/#hostpath-volume) when the `privileged` flag is enabled, but that one is marked as warning only (and it's probably unavoidable given the nature of the integration).

**What you expected to happen?**

>Please avoid saying "It should just work".

Well... it should ;) I should think changing the default value from 30 to 29 would be minimally disruptive and would save Digital Ocean users like myself an extra step.

**How to reproduce it?**

<details>
<summary>Here's my flux helm release entity:</summary>

```yaml
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: newrelic-bundle
spec:
  releaseName: bundle
  chart:
    spec:
      chart: nri-bundle
      sourceRef:
        kind: HelmRepository
        name: newrelic
        namespace: flux-system
      version: 3.5.0
  interval: 1h0m0s
  install:
    remediation:
      retries: 3
  # Default values
  # https://github.com/newrelic/helm-charts/blob/master/charts/nri-bundle/values.yaml
  values:
    # Default values for nri-infrastructure.
    # This is a YAML-formatted file.
    # Declare variables to be passed into your templates.
    # enables newrelic-infrastructure
    infrastructure:
      enabled: true
    # enables nri-prometheus
    prometheus:
      enabled: false
    # enables nri-metadata-injection
    webhook:
      enabled: true
    # enables kube-state-metrics
    ksm:
      enabled: true
    # enables nri-kube-events
    kubeEvents:
      enabled: true
    # enables newrelic-logging
    logging:
      enabled: false
    # enables newrelic-pixie
    newrelic-pixie:
      enabled: false
    # enables pixie-chart
    pixie-chart:
      enabled: false
    # enables newrelic-infra-operator
    newrelic-infra-operator:
      enabled: false
    # enables newrelic-k8s-metrics-adapter
    metrics-adapter:
      enabled: false
    # IMPORTANT: the following values are set as global so that they
    # can be shared by other newrelic product's charts
    #
    #   - Specify either the New Relic keys or the secret which
    #   contains them.
    #
    #   - Specify the Kubernetes cluster name.
    #     https://docs.newrelic.com/docs/kubernetes-monitoring-integration
    #
    global:
      lowDataMode: true
      licenseKey: ENC[REDACTED,type:str]
      #   insightsKey:
      #   customSecretName:
      #   customSecretLicenseKey:
      #   customSecretInsightsKey:
      cluster: REDACTED
    #   # IMPORTANT: Set this to true when deploying in EKS Fargate. This will pull the sidecar injector operator chart to
    #   # automatically monitor fargate workloads, and also prevent the DaemonSet from being scheduled in Fargate nodes
    #   fargate: true
    # It is possible to configure dependencies values from this chart.
    # For example you can configure `verboseLog` variable of the newrelic-infrastructure chart installed as a dependency and enable openshift configs as follows.
    newrelic-infrastructure:
      privileged: true
      resources:
        limits:
          cpu: 100m
          memory: 256Mi
        requests:
          cpu: 10m
          memory: 64Mi
```
</details>

You can reproduce my exact screenshot by installing the chart with these values on any [Digital Ocean Kubernetes (DOKS) cluster](https://www.digitalocean.com/products/kubernetes) and clicking on the operational readiness check in the Digital Ocean console.

Alternatively, you can run [clusterlint](https://github.com/digitalocean/clusterlint) directly... although this check is a DOKS-specific issue.

**Anything else we need to know?**

I know it's an easy fix, but happy to send a PR. Not sure it makes sense for me to sign a CLA for a one line change that I can easily work around on my side, hence the bug report.

**Comment by kang-makes** [comment]
Hi @ianonavy,

If I understood you correctly, the only issue you have is with the timeout of the MutatingWebhook.

The [timeout of the MutatingWebhook](https://github.com/newrelic/helm-charts/blob/master/charts/nri-metadata-injection/templates/admission-webhooks/mutatingWebhookConfiguration.yaml#L33) can be modified [changing your `values.yaml`](https://github.com/newrelic/helm-charts/blob/master/charts/nri-metadata-injection/values.yaml#L84).

Redeploying changing that value should be enough to cover your needs. Did that solve your issue?

**Comment by ianonavy** [comment]
@kang-makes Yes, I've already done that. Not looking for support—this is a product suggestion to change the default to 29.

**Comment by kang-makes** [comment]
With a little bit of investigation, I saw you are completely right. It is not an issue only with Digital Ocean but with every Kubernetes cluster

https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#timeouts
> The timeout value must be between 1 and 30 seconds.

However, we are moving these helm charts (and virtually as many charts as we can) to [the repository where the source code rests](https://github.com/newrelic/k8s-metadata-injection) and we are deprecating these charts. So yes, this is an issue, we have to solve it and if you want to create a PR solving it it should be done [here](https://github.com/newrelic/k8s-metadata-injection/blob/main/charts/nri-metadata-injection/values.yaml).

The charts are going to be changed during the next weeks.

We appreciate you taking the time to bring this to our attention.

**Comment by kang-makes** [comment]
Solved in newrelic/k8s-metadata-injection#171

Closing
